src/scene/light.js

```diff
@@ -156,6 +216,14 @@ class Light {
         return clone;
     }
 
+    get numCascades() {
+        return this.cascades.length;
```

Comment:
Doesn't look like the constructor creates this member? (and the clone function doesn't either).

Also, should it be named `_cascades`?

Reply:
both constructor and clone call setter on numCascades, which creates this member, so it is created.

Regarding the underscore, I think the rule here is that if there's a setter and getter for properties, the private variable name starts with _. If there isn't, it does not .. so from outside when you access them on light, you never use _.

Reply:
Ok you're relying on the setter to create the member. Of course this works, but I much prefer constructors to explicitly 'declare' their members (even setting them to null) for the sake of documentation and future maintenance. Also adding comments in the constructor (as you have done elsewhere) is extremely useful self-documentation. As it stands, I would suggest that looking at this code, it's much less obvious that light has a member called 'cascades'.

Regarding naming, the underscore is meant to indicate whether the property is internal or public (not whether it has an accessor) - same with function names. It can be a useful convention, though I don't think we follow it at all consistently in the engine.

Reply:
Yeah, this threw me a little bit too when I reviewed. I had to hunt for how it was defined, and while I was able to, it wasn't immediately obvious.

Reply:
I'll add the member. Just reading up on performance, it seems it's enough to assign null to a member in constructor to get performance benefit, and it does not need to be assigned consistent type. Note that I can find any conclusive evidence for this. What I was trying to do is to make sure an array is assigned to the member, instead of null and then array. But likely this makes no difference.

Reply:
and done.

Reply:
V8 can optimize objects when it is known for their properties in the constructor. If a number of properties change dynamically for an object, then this optimization might not occur, or it will go through a re-optimization process.
So attaching `null` in constructor - is already good, indeed.

Reply:
yep. In this case nothing has changed as constructor was calling a setter that was setting it to correct value .. so this is just for clarity when reading the code.
In general this is similar to few places where part of constructor is moved to some init function and called from constructor .. I would not initialize members to null before calling that init function.